src/audio/dp_queue.c

```diff
@@ -0,0 +1,300 @@
+	if (free_size) {
+		/* data consumed, free buffer space, no need for any special cache operations */
+		dp_queue->_read_offset =
+				dp_queue_inc_offset(dp_queue, dp_queue->_read_offset, free_size);
+	}
+
+	return 0;
+}
+
+static int dp_queue_set_ipc_params(struct dp_queue *dp_queue,
+				   struct sof_ipc_stream_params *params,
+				   bool force_update)
+{
+	CORE_CHECK_STRUCT(dp_queue);
+	if (dp_queue->_hw_params_configured && !force_update)
+		return 0;
+
+	dp_queue->audio_stream_params.frame_fmt = params->frame_fmt;
+	dp_queue->audio_stream_params.rate = params->rate;
+	dp_queue->audio_stream_params.channels = params->channels;
+	dp_queue->audio_stream_params.buffer_fmt = params->buffer_fmt;
+
+	dp_queue->_hw_params_configured = true;
+
+	return 0;
+}
+
+static int dp_queue_set_ipc_params_source(struct sof_source *source,
+					  struct sof_ipc_stream_params *params,
+					  bool force_update)
+{
+	struct dp_queue *dp_queue = dp_queue_from_source(source);
+
+	CORE_CHECK_STRUCT(dp_queue);
+	return dp_queue_set_ipc_params(dp_queue, params, force_update);
+}
+
+static int dp_queue_set_ipc_params_sink(struct sof_sink *sink,
+					struct sof_ipc_stream_params *params,
+					bool force_update)
+{
+	struct dp_queue *dp_queue = dp_queue_from_sink(sink);
+
+	CORE_CHECK_STRUCT(dp_queue);
+	return dp_queue_set_ipc_params(dp_queue, params, force_update);
+}
+
+static const struct source_ops dp_queue_source_ops = {
+	.get_data_available = dp_queue_get_data_available,
+	.get_data = dp_queue_get_data,
+	.release_data = dp_queue_release_data,
+	.audio_set_ipc_params = dp_queue_set_ipc_params_source,
+};
+
+static const struct sink_ops dp_queue_sink_ops = {
+	.get_free_size = dp_queue_get_free_size,
+	.get_buffer = dp_queue_get_buffer,
+	.commit_buffer = dp_queue_commit_buffer,
+	.audio_set_ipc_params = dp_queue_set_ipc_params_sink,
+};
+
+struct dp_queue *dp_queue_create(size_t ibs, size_t obs, uint32_t flags)
+{
+	struct dp_queue *dp_queue;
+
+	/* allocate DP structure */
+	if (flags & DP_QUEUE_MODE_SHARED)
+		dp_queue = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+				   sizeof(*dp_queue));
+	else
+		dp_queue = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*dp_queue));
+	if (!dp_queue)
+		return NULL;
+
+	dp_queue->_flags = flags;
+
+	CORE_CHECK_STRUCT_INIT(dp_queue, flags & DP_QUEUE_MODE_SHARED);
+
+	/* initiate sink/source */
+	source_init(dp_queue_get_source(dp_queue), &dp_queue_source_ops,
+		    &dp_queue->audio_stream_params);
+	sink_init(dp_queue_get_sink(dp_queue), &dp_queue_sink_ops,
+		  &dp_queue->audio_stream_params);
+
+	/* set obs/ibs in sink/source interfaces */
+	sink_set_obs(&dp_queue->_sink_api, obs);
+	source_set_ibs(&dp_queue->_source_api, ibs);
+
+	uint32_t max_ibs_obs = MAX(ibs, obs);
+	uint32_t min_ibs_obs = MIN(ibs, obs);
+
+	/* calculate required buffer size */
+	if (max_ibs_obs % min_ibs_obs == 0)
+		dp_queue->data_buffer_size = 2 * max_ibs_obs;
+	else
+		dp_queue->data_buffer_size = 3 * max_ibs_obs;
```

Comment:
I still don't understand this calculation...

Reply:
lest go with examples:

EXAMPLE 1:
 Consume 2 pieces each cycle
 produce 4 pieces every 2 cycles

buf size: 8

CYCLE0 prodice 4 pieces
buffer contains: 1111

CYCLE1 consume 2 pieces
buf: 11

CYCLE2 - we DO NOT know who's gonna be first - consumer or producer
  consumer first: empty -> 2222
  producer first: 112222 -> 2222
  
top buffer occupation is 6 ( 2*MAX(IBS,OBS)  = 8, operation safe)

EXAMPLE 2:
 Consume 3 pieces each 3 cycles
 produce 5 pieces every 5 cycles

(this cannot happen now, as currently there's no DP-DP connection, but may in the future)

cycle0 produce 3 pieces
111

cycle3 produce 3 pieces
111222

cycle5 - consumer must wait, not enough data
111222

cycle6 produce 3 pieces
111222333

cycle9 produce 3 pieces
111222333444 

cycle 10 - consume 5 pieces
2333444 

**top buffer occupation is 12.** 
2*MAX(IBS,OBS) = 10 - DROP!!!! no buffer space
3*MAX(IBS,OBS) = 15 operation safe